### ISSUE TITLE
[JENKINS-40619] Visualizing (list of sequential steps) as groups, as part of one stage

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3557]
<p>The pipeline view of Blue Ocean makes a great deal in listing a block of parallel steps by just displaying different boxes under each other. This seems to work only for parallel steps defined in one stage. If I add different build steps to a stage, operated sequentielly, the visualizing only displys one box for the entire stage. It would be helpful to express a group of steps, as part of the stage, processed sequentially, but shown as dedicated boxed (one group of steps in one box, the stage comprises a set of sequential groups of sequential steps).</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelhuettermann.png" title="michaelhuettermann's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelhuettermann">michaelhuettermann</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40619" target="_blank">Visualizing (list of sequential steps) as groups, as part of one stage</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: blueocean, parallel, sequential, stage, step, ui
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-01-04T01:53:50+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The pipeline view of Blue Ocean makes a great deal in listing a block of parallel steps by just displaying different boxes under each other. This seems to work only for parallel steps defined in one stage. If I add different build steps to a stage, operated sequentielly, the visualizing only displys one box for the entire stage. It would be helpful to express a group of steps, as part of the stage, processed sequentially, but shown as dedicated boxed (one group of steps in one box, the stage comprises a set of sequential groups of sequential steps).</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins ver. 2.37<br/>
Built at 16th December 2016 01:57 AM &#xB7; (no branch) &#xB7; 00384c1<br/>
Firefox 50.1.0
```
</details>
<details><summary><i>1 attachment</i></summary>

- [parallel.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35299/parallel.jpg)
  > ![parallel.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35299/parallel.jpg)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40619?focusedId=281511&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281511">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelhuettermann" class="user-hover" rel="michaelhuettermann">michaelhuettermann</a> I think this is related to <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> but could you please provide an example <tt>Jenkinsfile</tt> pipeline script and screenshot of how it is currently rendered to help clarify your request?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelhuettermann" class="user-hover" rel="michaelhuettermann">michaelhuettermann</a> I think this is related to <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> but could you please provide an example <tt>Jenkinsfile</tt> pipeline script and screenshot of how it is currently rendered to help clarify your request?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelhuettermann.png" title="michaelhuettermann's avatar" /> michaelhuettermann:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40619?focusedId=281684&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281684">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here a concrete example. The following snippet defines four tasks which are processed in parallel. In Blue Ocean pipeline view, this is wonderfully displayed as four boxes one after the other, see screenshot.</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
   stage 'Check preconditions'
   parallel "Sanity check 1": {
        echo "1"
    }, "Sanity check 2": {
        echo "2i"
    }, "Sanity check 3": {
        echo "3"
    }, "Sanity check 4": {
        echo "4"
    }
</pre>
</div></div>

<p>The idea now is to define the four example steps as a sequence of steps which have to be processed sequentially and still displayed as four paths as part of the stage (similar to the four parallel paths displayed now when run in parallel).</p>

<p>The current solution does not allow that. If I now define sequential steps as part of a stage, the single steps are displayed in the lower part of pipeline view, but the upper part of the pipeline view just shows one graphical control i.e. the stage (which sums up all steps). This is fine, but not as powerful as the requested functionality.</p>

<p>With other words, such a notation is requested:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
   stage 'Check preconditions'
   sequential "Sanity check 1": {
        echo "1"
    }, "Sanity check 2": {
        echo "2"
    }, "Sanity check 3": {
        echo "3"
    }, "Sanity check 4": {
        echo "4"
    }
</pre>
</div></div>

<p>This would lead to a very similar graphical presentation as the parallel processing, see attachment, but the difference is, that the four parts are run sequentially, not in parallel. It seems to be an implementation detail whether a key word <tt>sequential</tt> is used or something like <tt>group</tt>. Looks like <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> offers a generic approach to much more functionality than what is requested here.</p>
</pre>
</details>

</ul>
</details>

<p>Here a concrete example. The following snippet defines four tasks which are processed in parallel. In Blue Ocean pipeline view, this is wonderfully displayed as four boxes one after the other, see screenshot.</p>


```
 
   stage 'Check preconditions'
   parallel "Sanity check 1": {
echo "1"
    }, "Sanity check 2": {
echo "2i"
    }, "Sanity check 3": {
echo "3"
    }, "Sanity check 4": {
echo "4"
    }

```

<p>The idea now is to define the four example steps as a sequence of steps which have to be processed sequentially and still displayed as four paths as part of the stage (similar to the four parallel paths displayed now when run in parallel).</p>

<p>The current solution does not allow that. If I now define sequential steps as part of a stage, the single steps are displayed in the lower part of pipeline view, but the upper part of the pipeline view just shows one graphical control i.e. the stage (which sums up all steps). This is fine, but not as powerful as the requested functionality.</p>

<p>With other words, such a notation is requested:</p>


```
 
   stage 'Check preconditions'
   sequential "Sanity check 1": {
echo "1"
    }, "Sanity check 2": {
echo "2"
    }, "Sanity check 3": {
echo "3"
    }, "Sanity check 4": {
echo "4"
    }

```

<p>This would lead to a very similar graphical presentation as the parallel processing, see attachment, but the difference is, that the four parts are run sequentially, not in parallel. It seems to be an implementation detail whether a key word <tt>sequential</tt> is used or something like <tt>group</tt>. Looks like <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> offers a generic approach to much more functionality than what is requested here.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40619?focusedId=281759&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281759">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Right that makes sense to me. Thanks for reporting this!</p>
</pre>
</details>

</ul>
</details>

<p>Right that makes sense to me. Thanks for reporting this!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40619?focusedId=281963&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281963">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>On further inspection this seems like a use case that could be covered by nested stages. Closing this ticket as a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> but we will review these and other linked tickets when cooking up a solution.</p>
</pre>
</details>

</ul>
</details>

<p>On further inspection this seems like a use case that could be covered by nested stages. Closing this ticket as a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> but we will review these and other linked tickets when cooking up a solution.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-38442%22">JENKINS-38442</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35750%22">JENKINS-35750</a>]</i>
